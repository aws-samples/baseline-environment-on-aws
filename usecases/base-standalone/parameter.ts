import { Environment } from 'aws-cdk-lib';

export interface MyParameter {
  env?: Environment;
  envName: string;
  securityNotifyEmail: string;
  chatbotSlackWorkspaceId: string;
  chatbotSlackChannelId: string;
}

// Example
export const DevParameter: MyParameter = {
  envName: 'Development',
  securityNotifyEmail: 'notify-security@example.com',
  chatbotSlackWorkspaceId: 'T8XXXXXXX',
  chatbotSlackChannelId: 'C00XXXXXXXX',
};
