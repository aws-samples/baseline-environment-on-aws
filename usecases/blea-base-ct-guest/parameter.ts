import { Environment } from 'aws-cdk-lib';

export interface MyParameter {
  env?: Environment;
  envName: string;
  securityNotifyEmail: string;

  // (optional) AWS CodeStar Connections parameters for CDK Pipelines.
  // Only used in bin/blea-base-ct-guest-via-cdk-pipelines.ts
  sourceRepository?: string;
  sourceBranch?: string;
  sourceConnectionArn?: string;
}

// Example
export const DevParameter: MyParameter = {
  envName: 'Development',
  securityNotifyEmail: 'notify-security@example.com',
  sourceRepository: 'aws-samples/baseline-environment-on-aws',
  sourceBranch: 'main',
  sourceConnectionArn: 'arn:aws:codestar-connections:ap-northeast-1:xxxxxxxxxxxx:connection/example',
};
