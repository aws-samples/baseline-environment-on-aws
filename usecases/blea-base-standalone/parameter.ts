import { Environment } from 'aws-cdk-lib';

export interface MyParameter {
  env?: Environment;
  envName: string;
  securityNotifyEmail: string;
}

// Example
export const DevParameter: MyParameter = {
  envName: 'Development',
  securityNotifyEmail: 'notify-security@example.com',
  // env: { account: '123456789012', region: 'ap-northeast-1' },
};
