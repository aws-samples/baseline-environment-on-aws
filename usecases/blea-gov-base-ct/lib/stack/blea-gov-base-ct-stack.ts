import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Iam } from '../construct/iam';
import { Logging } from '../construct/logging';
import { Detection } from '../construct/detection';
import { Notification } from '../construct/notification';

export interface BLEAGovBaseCtStackProps extends StackProps {
  securityNotifyEmail: string;
  securitySlackWorkspaceId?: string;
  securitySlackChannelId?: string;
}

export class BLEAGovBaseCtStack extends Stack {
  constructor(scope: Construct, id: string, props: BLEAGovBaseCtStackProps) {
    super(scope, id, props);

    new Iam(this, 'Iam');

    // AWS CloudTrail configuration in Control Tower Landing Zone v3.0 will not create CloudWatch Logs LogGroup in each Guest Accounts.
    // And it will delete these LogGroups when AWS CloudTrial Configuration is disabled in case of updating Landing Zone version from older one.
    // BLEA should notify their alarms continuously. So, if there is no CloudTrail and CloudWatch Logs in Guest Account, BLEA creates them to notify the Alarms.
    const logging = new Logging(this, 'Logging');

    // Security Alarms
    // !!! Need to setup SecurityHub, GuardDuty manually on Organizations Management account
    // AWS Config and CloudTrail are set up by Control Tower
    const detection = new Detection(this, 'Detection', {
      notifyEmail: props.securityNotifyEmail,
      cloudTrailLogGroupName: logging.trailLogGroup.logGroupName,
    });

    if (!props.securitySlackWorkspaceId || !props.securitySlackChannelId) {
      throw new Error('securitySlackWorkspaceId and securitySlackChannelId are required');
    }

    new Notification(this, 'Notification', {
      topicArn: detection.topic.topicArn,
      workspaceId: props.securitySlackWorkspaceId,
      channelId: props.securitySlackChannelId,
    });
  }
}
