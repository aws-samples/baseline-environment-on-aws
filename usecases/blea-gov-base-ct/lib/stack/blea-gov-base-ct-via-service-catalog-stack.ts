import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { CloudFormationProduct, Portfolio, ProductStack, ProductStackHistory } from 'aws-cdk-lib/aws-servicecatalog';
import { Iam } from '../construct/iam';
import { Logging } from '../construct/logging';
import { Detection } from '../construct/detection';
import { Role } from 'aws-cdk-lib/aws-iam';

export interface BLEAGovBaseCtScStackProps extends StackProps {
  securityNotifyEmail: string;
}

export class BLEAGovBaseCtScStack extends Stack {
  constructor(scope: Construct, id: string, props: BLEAGovBaseCtScStackProps) {
    super(scope, id, props);

    class BaselineProduct extends ProductStack {
      constructor(scope: Construct, id: string) {
        super(scope, id);

        new Iam(this, 'Iam');

        // AWS CloudTrail configuration in Control Tower Landing Zone v3.0 will not create CloudWatch Logs LogGroup in each Guest Accounts.
        // And it will delete these LogGroups when AWS CloudTrial Configuration is disabled in case of updating Landing Zone version from older one.
        // BLEA should notify their alarms continuously. So, if there is no CloudTrail and CloudWatch Logs in Guest Account, BLEA creates them to notify the Alarms.
        const logging = new Logging(this, 'Logging');

        // Security Alarms
        // !!! Need to setup SecurityHub, GuardDuty manually on Organizations Management account
        // AWS Config and CloudTrail are set up by Control Tower
        new Detection(this, 'Detection', {
          notifyEmail: props.securityNotifyEmail,
          cloudTrailLogGroupName: logging.trailLogGroup.logGroupName,
        });
      }
    }

    const portfolio = new Portfolio(this, 'Portfolio', {
      displayName: 'BLEA Baselines for Guest accounts',
      providerName: 'Platform team at Example Company',
    });

    // Use history and lock each versions to update blueprints from AWS Control Tower.
    // See: https://docs.aws.amazon.com/controltower/latest/userguide/update-a-blueprint.html
    const productStackHistory = new ProductStackHistory(this, 'ProductStackHistory', {
      productStack: new BaselineProduct(this, 'BaselineProduct'),
      currentVersionName: 'v1',
      currentVersionLocked: true,
    });

    const product = new CloudFormationProduct(this, 'Product', {
      productName: 'BLEA Baseline',
      owner: 'Platform team at Example Company',
      productVersions: [productStackHistory.currentVersion()],
    });

    portfolio.addProduct(product);

    // Allow access to AWSControlTowerExecution role for debugging
    portfolio.giveAccessToRole(Role.fromRoleName(this, 'CTRole', 'AWSControlTowerExecution'));
  }
}
