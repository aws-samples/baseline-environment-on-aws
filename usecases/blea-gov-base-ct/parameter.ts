import { Environment } from 'aws-cdk-lib';

export interface AppParameter {
  env?: Environment;
  envName: string;
  securityNotifyEmail: string;
  securitySlackWorkspaceId?: string; // required if deploy via CLI
  securitySlackChannelId?: string; // required if deploy via CLI
}

export interface PipelineParameter {
  env: Environment;
  envName: string;

  // AWS CodeStar Connections parameters for CDK Pipelines.
  // Only used in bin/blea-gov-base-ct-via-cdk-pipelines.ts
  sourceRepository: string;
  sourceBranch: string;
  sourceConnectionArn: string;
}

// Example for Development
export const devParameter: AppParameter = {
  envName: 'Development',
  securityNotifyEmail: 'notify-security@example.com',
  securitySlackWorkspaceId: 'T8XXXXXXX',
  securitySlackChannelId: 'C00XXXXXXXX',
  // env: { account: '123456789012', region: 'ap-northeast-1' },
};

// Example for Staging
export const stagingParameter: AppParameter = {
  envName: 'Staging',
  securityNotifyEmail: 'notify-security@example.com',
  env: { account: '123456789012', region: 'ap-northeast-1' },
};

// Example for Pipeline Deployment
export const devPipelineParameter: PipelineParameter = {
  env: { account: '123456789012', region: 'ap-northeast-1' },
  envName: 'DevPipeline',
  sourceRepository: 'aws-samples/baseline-environment-on-aws',
  sourceBranch: 'main',
  sourceConnectionArn: 'arn:aws:codestar-connections:ap-northeast-1:xxxxxxxxxxxx:connection/example',
};
