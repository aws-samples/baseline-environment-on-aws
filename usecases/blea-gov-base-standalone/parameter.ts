import { Environment } from 'aws-cdk-lib';

export interface AppParameter {
  env?: Environment;
  envName: string;
  securityNotifyEmail: string;
  securitySlackWorkspaceId: string; // required if deploy via CLI
  securitySlackChannelId: string; // required if deploy via CLI
}

// Example
export const devParameter: AppParameter = {
  envName: 'Development',
  securityNotifyEmail: 'notify-security@example.com',
  securitySlackWorkspaceId: 'T8XXXXXXX',
  securitySlackChannelId: 'C00XXXXXXXX',
  // env: { account: '123456789012', region: 'ap-northeast-1' },
};
