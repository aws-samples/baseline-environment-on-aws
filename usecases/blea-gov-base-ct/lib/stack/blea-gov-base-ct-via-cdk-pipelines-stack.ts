import * as cdk from 'aws-cdk-lib';
import { pipelines, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { BLEAGovBaseCtStage } from '../stage/blea-gov-base-ct-stage';
import { AppParameter } from '../../parameter';

export interface BLEAGovBaseCtPipelineStackProps extends cdk.StackProps {
  targetParameters: AppParameter[];
  sourceRepository: string;
  sourceBranch: string;
  sourceConnectionArn: string;
}

export class BLEAGovBaseCtPipelineStack extends Stack {
  constructor(scope: Construct, id: string, props: BLEAGovBaseCtPipelineStackProps) {
    super(scope, id, props);

    const pipeline = new pipelines.CodePipeline(this, 'Pipeline', {
      synth: new pipelines.ShellStep('Synth', {
        input: pipelines.CodePipelineSource.connection(props.sourceRepository, props.sourceBranch, {
          connectionArn: props.sourceConnectionArn,
        }),
        installCommands: ['n stable', 'node --version', 'npm i -g npm', 'npm --version'],
        commands: [
          'npm ci --workspaces',
          'cd usecases/blea-gov-base-ct',
          'npx cdk synth --app "npx ts-node --prefer-ts-exts bin/blea-gov-base-ct-via-cdk-pipelines.ts" --all',
        ],
        primaryOutputDirectory: './usecases/blea-gov-base-ct/cdk.out',
      }),
    });

    props.targetParameters.forEach((params) => {
      pipeline.addStage(new BLEAGovBaseCtStage(this, 'Dev', params));
    });
  }
}
