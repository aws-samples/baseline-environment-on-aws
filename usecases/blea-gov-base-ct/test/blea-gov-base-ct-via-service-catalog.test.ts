import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { devParameter } from '../parameter';
import { BLEAGovBaseCtScStack } from '../lib/stack/blea-gov-base-ct-via-service-catalog-stack';

test('Snapshot test for BLEGovABase Stack', () => {
  const app = new cdk.App();
  const stack = new BLEAGovBaseCtScStack(app, 'Dev-BLEAGovBaseCtSc', {
    env: devParameter.env,
    tags: {
      Repository: 'aws-samples/baseline-environment-on-aws',
      Environment: devParameter.envName,
    },

    securityNotifyEmail: devParameter.securityNotifyEmail,
  });

  // test with snapshot
  expect(Template.fromStack(stack)).toMatchSnapshot();
});
