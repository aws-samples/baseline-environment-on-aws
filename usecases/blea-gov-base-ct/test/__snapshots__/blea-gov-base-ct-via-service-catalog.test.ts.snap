// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for BLEGovABase Stack 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Portfolio856A4190": {
      "Properties": {
        "DisplayName": "BLEA Baselines for Guest accounts",
        "ProviderName": "Platform team at Example Company",
      },
      "Type": "AWS::ServiceCatalog::Portfolio",
    },
    "PortfolioPortfolioProductAssociationd2b2e10d104073565C5B": {
      "Properties": {
        "PortfolioId": {
          "Ref": "Portfolio856A4190",
        },
        "ProductId": {
          "Ref": "Product896941B4",
        },
      },
      "Type": "AWS::ServiceCatalog::PortfolioProductAssociation",
    },
    "PortfolioPortolioPrincipalAssociationd82fda574793DAB993CA": {
      "Properties": {
        "PortfolioId": {
          "Ref": "Portfolio856A4190",
        },
        "PrincipalARN": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":iam::",
              {
                "Ref": "AWS::AccountId",
              },
              ":role/AWSControlTowerExecution",
            ],
          ],
        },
        "PrincipalType": "IAM",
      },
      "Type": "AWS::ServiceCatalog::PortfolioPrincipalAssociation",
    },
    "Product896941B4": {
      "Properties": {
        "Name": "BLEA Baseline",
        "Owner": "Platform team at Example Company",
        "ProvisioningArtifactParameters": [
          {
            "DisableTemplateValidation": false,
            "Info": {
              "LoadTemplateFromURL": {
                "Fn::Sub": "https://s3.\${AWS::Region}.\${AWS::URLSuffix}/cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}/HASH-REPLACED.json",
              },
            },
            "Name": "v1",
          },
        ],
      },
      "Type": "AWS::ServiceCatalog::CloudFormationProduct",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
